### ISSUE TITLE
Size mismatch error after one round of federated training with Flower and YoloV8

### ISSUE THREAD

**Issue by shubhamjanu** [ultralytics/ultralytics#9554]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello Ultralytics Team!

I am working on a federated learning project for my Master's research. Federated training completes for one round and when parameters are received for second round of training. set_parameter() method throws error related to size mismatch of tensors. I have attached my code and error log. Please suggest how this issue can be resolved.
[client.txt](https://github.com/ultralytics/ultralytics/files/14873278/client.txt)
[log.txt](https://github.com/ultralytics/ultralytics/files/14873279/log.txt)

**Comment by glenn-jocher** [comment]
@shubhamjanu hello! Thanks for reaching out with details about your federated learning project. It's great to hear you're diving into such an interesting area with YOLOv8. 😊

The tensor size mismatch error you're encountering during the second round of federated training usually indicates a difference in the expected model architecture or input dimensions between the two rounds. Without diving into your specific files, a common cause for this might be changes in the model architecture definitions or inconsistencies in data preprocessing across rounds.

To address the issue, ensure all clients in the federated learning setup use the **exact same model architecture** and **data preprocessing steps**. Here's a quick checklist:
- Verify model loading processes to ensure consistency in model architecture across training rounds.
- Ensure that input sizes (`imgsz`) and model configurations are identical when initializing the model for each round.
- Double-check that the tensors being passed around match the expected dimensions on both sender and receiver sides.

As a starting point, you can compare the model's expected input size and architecture between rounds by inspecting your model initialization and parameter update code snippets, ensuring they align perfectly.

If you continue to face issues, it might help to provide a minimal reproducible example that showcases how the clients and server exchange data and update model parameters. This could offer more insight into where the mismatch might be occurring.

Feel free to share additional snippets or findings based on these suggestions. We're here to help!

**Comment by shubhamjanu** [comment]
Hi @glenn-jocher , Initially I am loading Yolov8m.pt model for every client as you can see in my code. I have only two clients which get data_path as japan0.yaml and japan1.yaml. Imgsz is also defined constantly at 640. I am not able to understand why there is a mismatch in second round. I even flagged the issue on Flower official slack channel but they are not so familiar with ultralytics. One more thing to highlight here is: when i run federated code using start_server() method, the training runs fine. The only problem with it was I was unable to utilize my single GPU so i had to switch to start_simulation() method. The basic structure of the code is almost same. But, with start_simulation(), it gives size mismatch error. Could you please suggest something?

**Comment by glenn-jocher** [comment]
Hi @shubhamjanu 👋, thanks for the detailed info! It seems like you've been thorough with keeping the model and input sizes consistent across clients, which is great! The issue might lie in how the model's parameters are being updated or handled differently between `start_server()` and `start_simulation()` methods. 

A common cause for a size mismatch error, especially when switching methods like you described, could be the way tensors or model weights are serialized and deserialized between rounds. This might inadvertently alter the shape or size of the parameters being passed back to the model.

My suggestion would be to double-check:
- The serialization and deserialization process (if any) applied to the model's state_dict or parameters when being sent over the network. Ensure consistency in how parameters are packaged and unpacked.
- How the model is re-initialized or updated with new parameters in the `start_simulation()` method. Verify that the model state is correctly and fully restored with no missing or extra parameters.

As a starting point, you can log the shape of key parameters before sending and after receiving to pinpoint any discrepancy. Something like:

```python
# Before sending
for param in model.state_dict().values():
    print(param.size())

# After receiving and before updating the model
for param in received_state_dict.values():
    print(param.size())
```

This can help verify that the shapes match up exactly before and after the transfer. If you find discrepancies, it might hint at where in the process things are getting altered.

Hope this points you in the right direction! Let us know how it goes. 🚀

**Comment by LvAoping** [comment]
Hi @shubhamjanu. I had the similar issue while implementing flower framework with yolov8. It works well for the ROUND 1, 
`INFO :      configure_fit: strategy sampled 2 clients (out of 2)`
`INFO :      aggregate_fit: received 2 results and 0 failures`
`INFO :      configure_evaluate: strategy sampled 2 clients (out of 2)`
`INFO :      aggregate_evaluate: received 2 results and 0 failures`,
However, it gave a lot size mismatch errors when copying the weight and bais from each layer. 
Have you solved this problem? Could you give some more suggestions?

**Comment by glenn-jocher** [comment]
Hi @shubhamjanu and @shubhamjanu 👋! The behavior you're observing might be due to differences in model state initialization between the base `yolov8l.pt` and a `last.pt` that's been trained for at least one epoch. When using a pre-trained model like `yolov8l.pt`, the model's parameters are set for a specific architecture and input size. If your federated learning setup slightly differs (e.g., different input size or custom layers), initializing with `yolov8l.pt` might lead to mismatch errors. 

On the other hand, `last.pt` contains parameters adjusted to your specific setup during the initial training round. So, initializing with `last.pt` ensures compatibility with your federated learning architecture, bypassing the mismatch issue.

A simple workaround could be starting your federated learning with a round of training on a small subset of your data to generate a compatible `last.pt`, and then proceeding with the federated training setup.

```python
# Example: Train briefly to generate a compatible 'last.pt'
model.train(data='small_subset.yaml', epochs=1)
```

Keep in mind, ensuring the model architecture and input sizes are consistent across your federated setup is key! Hope this helps! 😊

**Comment by shubhamjanu** [comment]
@GTziolas Hi George !
When I run one epoch on my dataset, I obtain best.pt and last.pt, Now, you can use this best.pt/last.pt  in federated learning setup by simply replacing yolov8l.pt with best.pt or last.pt

**Comment by DARRICK1103** [comment]
Hi @shubhamjanu , I just ran your code. Do you have any ideas on how to save the final aggregated YOLOv8 model?

**Comment by shubhamjanu** [comment]
@DARRICK1103 Please see my evaluate method. You can save model using torch.save in evaluate method.
def evaluate(self, parameters, config):
        round_num = config.get("server_round", 0)  # Get the current round number from the config
        print("i am starting evaluation")
        set_parameters([self.net](http://self.net/), parameters)
        torch.save([self.net](http://self.net/),f'FL_round_{round_num}.pt')
        print("The size of testing data is: ",self.testLength)
        mAP50 = test([self.net](http://self.net/),data=self.data_path,imgsz=self.imgsz)
        save_map50_value(round_num, int(self.client_id), mAP50)
        return float(mAP50),self.testLength,{"mAP50": mAP50}

**Comment by DARRICK1103** [comment]
@shubhamjanu Thanks for the quick response! However, as I understand it, the evaluate method in the FlowerClient class is intended to evaluate the client model, not the aggregated global model. So, I assume it is saving the client model, not the final aggregated model using FedAvg.

Please correct me if I'm wrong.

**Comment by glenn-jocher** [comment]
Hi @DARRICK1103! You're correct in your understanding. The `evaluate` method in the `FlowerClient` is indeed for evaluating the client model, not the final aggregated global model. Your approach to saving the aggregated model using the `SaveModelStrategy` is spot on for capturing the global model state after each round of federated averaging.

Regarding the warning you encountered, it's recommended to use the model's specific save method if available, as it might handle certain aspects of the model's state more appropriately than `torch.save()`. For YOLOv8 models managed with Ultralytics, using `model.save('filename.pt')` ensures that all components of the model are correctly serialized, especially if there are custom layers or specific state elements that `torch.save()` might not capture effectively.

Here's a quick adjustment to ensure you're using the best practice for saving YOLO models:

```python
def save_aggregated_model(parameters: List[np.ndarray], round_num: int):
    model = load_model()
    set_parameters(model, parameters)
    model.save(f'model/FL_aggregated_round_{round_num}.pt')  # Using model's save method
```

This should help mitigate the warning and ensure your model is saved in the best format for later use. Keep up the great work with federated learning! 🚀

**Comment by shubhamjanu** [comment]
@DARRICK1103 @glenn-jocher Hi Glenn ! The model which I am saving under evaluate method is the aggregated model itself. I have verified it. Also, the Flower documentation mentions that this evaluate method is client-side evaluation of aggregated model. Please refer https://flower.ai/docs/framework/tutorial-series-use-a-federated-learning-strategy-pytorch.html

**Comment by glenn-jocher** [comment]
Hi @shubhamjanu! Thanks for the clarification and for sharing the Flower documentation reference. It's great to hear that you've verified the model saved in the `evaluate` method is indeed the aggregated model. This insight is valuable for ensuring that everyone understands the workflow correctly. Keep up the excellent work, and feel free to reach out if you have more questions or insights as you progress with your project! 🌟

**Comment by shubhamjanu** [comment]
@glenn-jocher Hello Glenn! Are the YoloV9 and YoloV10 models integrated in Ultralytics package? I want to upgrade my federated learning code with latest yolo model (preferably YoloV10). If they are integrated already, How can I use them? If not integrated, when can i expect them to be integrated?

**Comment by glenn-jocher** [comment]
Hi @shubhamjanu! Currently, YOLOv9 and YOLOv10 are not integrated into the Ultralytics package. Our team is always working on integrating the latest advancements, but we don't have a specific timeline for these models yet. Keep an eye on our repository updates for any future announcements regarding new model integrations. Thanks for your interest in using the latest YOLO models! 🚀

**Comment by DARRICK1103** [comment]
Hi @glenn-jocher @shubhamjanu 
def train(net, data, epochs: int):
    net.train(data=data, epochs=epochs, imgsz=640, batch=4, plots=False, device=0)
When we do this training, will it cause the weight file to forget the old data since I retrain the pretrained model with the new data?

**Comment by glenn-jocher** [comment]
Hi @DARRICK1103,

When you retrain a pretrained model with new data, the model will update its weights based on the new data. This process is known as fine-tuning. The extent to which the model "forgets" the old data depends on several factors, including the size of the new dataset, the number of epochs, and the learning rate.

To mitigate the risk of catastrophic forgetting (where the model forgets previously learned information), you can:
1. Use a lower learning rate to ensure gradual updates to the weights.
2. Fine-tune only the later layers of the model by freezing the initial layers. This can be done by setting the `freeze` parameter in your training configuration.

Here's an example of how you can freeze the initial layers:
```python
net.train(data=data, epochs=epochs, imgsz=640, batch=4, plots=False, device=0, freeze=10)
```

This will freeze the first 10 layers of the model, allowing only the remaining layers to be updated during training.

Feel free to experiment with these settings to find the best balance for your specific use case. If you encounter any issues, please ensure you are using the latest version of the Ultralytics package and provide a reproducible example if the issue persists.

Best of luck with your training! 🚀
